### ISSUE TITLE
dub test fails due to missing tester module

### ISSUE THREAD

**Issue by joseph-wakeling-frequenz** [sociomantic-tsunami/ocean#716]
If we try to run `dub test` (which uses `dub.sdl` to manage the build), building the unittest executable fails with the message:

```
Error: module `tester` is in file 'tester.d' which cannot be read
```

It's not clear what is responsible here, since AFAICS there is nowhere in ocean (or its integration tests) that tries to `import tester`.

**Comment by mihails-strasuns** [comment]
It is a dub bug, it interprets https://github.com/sociomantic-tsunami/ocean/blob/e53ac93fbf3bfa9b2dceec1a2b6dc4a0ec7f78b2/src/ocean/core/UnitTestRunner.d#L16 as an actual module declaration.

**Comment by joseph-wakeling-frequenz** [comment]
That's grim :-\  Want me to report it upstream, or is someone already on it?

**Comment by matthias-wende-sociomantic** [comment]
It seems there is nothing we can do here. Can we close this issue?

**Comment by joseph-wakeling-frequenz** [comment]
@matthias-wende-sociomantic this may be an upstream bug but it should not be closed until it's reported upstream.  Even then I'd recommend leaving this issue open as a tracker for the upstream issue.

You _could_ go full WONTFIX but since the `dub.sdl` file is in place in the repo, I presume you want it to be usable with DUB, so it's probably worth caring about issues that prevent that ;-)

**Comment by mihails-strasuns** [comment]
You can workaround it by blacklisting test runner module in `dub.sdl` for unittest build - it may otherwise clash with default test runner anyway (not sure how it is currently implemented in dub).

**Comment by joseph-wakeling-frequenz** [comment]
I got a few minutes to boil this down to a minimal example, and reported the issue upstream: https://github.com/dlang/dub/issues/1761

The issue appears to be that the `module` statement in these ddoc blocks precedes the real `module` statement.  So a workaround could be to just avoid `module` statements in module ddoc banners.

**Comment by andrew-stevenson-sociomantic** [comment]
There is now a workaround in dub for this so maybe this can be closed?

**Comment by Geod24** [comment]
I think so too (although the work around will only be part of the 2.089 release cycle)

**Comment by joseph-wakeling-frequenz** [comment]
I've asked if we can get a point release to speed things up there.  The worst they can say is No :-P

What I suggest is that someone (I'm happy to do it) submits a PR to revert the workaround patch, which we block until a DUB release with the fix is out.  We can close this issue then, and just leave the open PR to remind us of the need for follow-up.

Once a DUB release with the fix is out, and verified to work with ocean, unblock and merge the PR.

It would probably also be helpful to add a CI job verifying that ocean works with DUB from there on out, but that's a different concern ;-)

**Comment by Geod24** [comment]
I'd rather not have the workaround reverted for the time being, as it costs nothing, but allow us to use Ocean with older compiler (and thus, dub) versions. A new `dub` version is not going to fix this.

**Comment by joseph-wakeling-frequenz** [comment]
> I'd rather not have the workaround reverted for the time being

That's why I suggested leaving the reversion PR open until an updated DUB is released (or even for a time window after that), but sure.  There is no rush on that.

**Comment by joseph-wakeling-frequenz** [comment]
Anyway, just to verify, I did a fresh build of `dub` from master (including @andrew-stevenson-sociomantic's fix), and then attempted to run `dub test` on ocean v5.1.1.

If I specify `--compiler=dmd-transitional`, I get the following results:

```
Generating test runner configuration 'ocean-test-library' for 'library' (sourceLibrary).
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Performing "unittest" build using dmd-transitional for x86_64.
ocean 5.1.1: building configuration "ocean-test-library"...
/tmp/dub_test_root_9e8fa4f7_f0c1_492d_81b2_73f3fd223b98.d(2,8): Error: module typetuple is in file 'std/typetuple.d' which cannot be read
import path[0] = src
import path[1] = /usr/include/d2/dmd-transitional
dmd-transitional failed with exit code 1.
```

I'm assuming this is down to absence of phobos, because when I use `--compiler=dmd` instead I get a different set of errors:

```
Generating test runner configuration 'ocean-test-library' for 'library' (sourceLibrary).
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Performing "unittest" build using dmd for x86_64.
ocean 5.1.1: building configuration "ocean-test-library"...
There will be no gc stats! You need to use dmd transitional:
https://github.com/sociomantic-tsunami/dmd-transitional
src/ocean/application/components/ConfigOverrides.d(176,18): Error: undefined identifier NamedTest
src/ocean/application/components/Version.d(134,5): Error: template instance `test!"=="` template test is not defined
src/ocean/application/components/Version.d(138,5): Error: template instance `test!"=="` template test is not defined
src/ocean/application/components/Version.d(149,5): Error: template instance `test!"=="` template test is not defined
src/ocean/core/Array.d(114,10): Error: undefined identifier concat_test
src/ocean/core/Array.d(122,10): Error: undefined identifier concat_test
src/ocean/core/Array.d(123,10): Error: undefined identifier concat_test
src/ocean/core/Array.d(128,10): Error: undefined identifier concat_test
src/ocean/core/Array.d(133,10): Error: undefined identifier concat_test
src/ocean/core/BitArray.d(144,9): Error: undefined identifier test
src/ocean/core/BitArray.d(147,13): Error: undefined identifier test
src/ocean/core/BitArray.d(225,9): Error: undefined identifier test
src/ocean/core/BitArray.d(226,9): Error: undefined identifier test
src/ocean/core/BitArray.d(227,9): Error: undefined identifier test
src/ocean/core/BitArray.d(228,9): Error: undefined identifier test
src/ocean/core/BitArray.d(229,9): Error: undefined identifier test
src/ocean/core/BitArray.d(232,9): Error: undefined identifier test
src/ocean/core/BitArray.d(234,9): Error: undefined identifier test
src/ocean/core/BitArray.d(240,9): Error: undefined identifier test
src/ocean/core/BitArray.d(242,9): Error: undefined identifier test
dmd failed with exit code 1.
```

So @andrew-stevenson-sociomantic's fix takes things forwards but still leaves us with a number of extra things to address if we want to be able to fully use ocean with DUB.

**Comment by joseph-wakeling-frequenz** [comment]
I was using DMD 2.085.1 in the second of the above `dub test` runs, just for the record.

**Comment by joseph-wakeling-frequenz** [comment]
I think the `dmd` failures are because of `version (UnitTest)`.  Will try to confirm.

**Comment by joseph-wakeling-frequenz** [comment]
Yup, it's definitely down to `version (UnitTest)`.  If I run `dub test` on v5.x.x plus a revert of the workaround patch, all is fine, except that we come up against the `UnitTestRunner` `main` function.  If that's versioned out, then we run into linker errors (unsurprising, because `dub.sdl` specifies no libraries to link against).

TL;DR I think @andrew-stevenson-sociomantic's fix solves the main blocker on using ocean as a DUB dependency of another app.

**Comment by joseph-wakeling-frequenz** [comment]
This can be closed, as the latest dub release (1.18.0) fixes the problem, thanks to @andrew-stevenson-sociomantic's patch.

There are other blockers to testing ocean with DUB, but I'll open other issues for that ;-)

Thanks @andrew-stevenson-sociomantic for addressing the problem at its source!
